test/framework/src/main/java/org/elasticsearch/transport/nio/NioTransport.java

```diff
@@ -169,14 +176,48 @@ protected void stopInternal() {
     }
 
     protected SocketEventHandler getSocketEventHandler() {
-        return new SocketEventHandler(logger, this::exceptionCaught, openChannels);
+        return new SocketEventHandler(logger);
     }
 
     final void exceptionCaught(NioSocketChannel channel, Exception exception) {
-        onException(channel, exception);
+        onException((TcpNioSocketChannel) channel, exception);
     }
 
     private Consumer<NioSocketChannel> getContextSetter(String profileName) {
-        return (c) -> c.setContexts(new TcpReadContext(c, new TcpReadHandler(profileName,this)), new TcpWriteContext(c));
+        return (c) -> {
+            c.setContexts(new TcpReadContext(c, new TcpReadHandler(profileName,this)), new TcpWriteContext(c));
+            c.setExceptionHandler(this::exceptionCaught);
+        };
+    }
+
+    private void acceptChannel(NioSocketChannel channel) {
+        TcpNioSocketChannel tcpChannel = (TcpNioSocketChannel) channel;
+        openChannels.acceptedChannelOpened(tcpChannel);
+        tcpChannel.addCloseListener(new RemoveClosedChannelListener(channel));
```

Comment:
rather use `ActionListener.wrap(v -> openChannels.channelClosed(channel), e -> openChannels.channelClosed(channel))` then we don't need to define this extra class